### ISSUE TITLE
Error in Load model weights

### ISSUE THREAD

**Issue by inakanishi** [rmokady/CLIP_prefix_caption#76]
When I started ClipCap on google colab (Open in Colab), I got the following error at the Load model weights. How can I solve this problem?

<img width="806" alt="image" src="https://github.com/rmokady/CLIP_prefix_caption/assets/36018302/f0ab30f3-76dd-4a38-b79c-36e9e4d51301">

**Comment by abrar-fahim** [comment]
I did some digging and found this to be a version issue with the transformers module.

TLDR: Simply deleting the bias and masked_bias keys from the state dictionary seems to work. Something like:

```
altered_state_dict = torch.load(model_path, map_location=CPU)
for i in range(12):
    del altered_state_dict['gpt.transformer.h.' + str(i) + '.attn.bias']
    del altered_state_dict['gpt.transformer.h.' + str(i) + '.attn.masked_bias']
```

And then loading the altered state dictionary into the model by:

```
model.load_state_dict(altered_state_dict)
```

## Most likely explanation

More specifically, in transformers: version 4.11.3 used in the original notebook, in the source code for GPT2Attention, "bias" and "masked_bias" are persistent buffers, and therefore saved to the state_dictionary: 
![image](https://github.com/rmokady/CLIP_prefix_caption/assets/41694147/e31a0c19-5c9f-48aa-989a-6cd70eddda91)

In the latest transformers version (version 4.33.1),  in GPT2Attention, "bias" and "masked_bias" buffers are defined to have persistent=False, meaning that they are no longer a part of the model's state dictionary:
![image](https://github.com/rmokady/CLIP_prefix_caption/assets/41694147/dc73b79d-8b08-47f2-a09b-c6a06721a62d)

**Comment by inakanishi** [comment]
Thank you for your careful explanation. I did it!

**Comment by Hsn37** [comment]
An easier workaround for this is to set `strict=False` in the `load_state_dict` function.

```
model.load_state_dict(model_weights, strict=False) 
```

**Comment by returnaaa** [comment]
> I did some digging and found this to be a version issue with the transformers module.
> 
> TLDR: Simply deleting the bias and masked_bias keys from the state dictionary seems to work. Something like:
> 
> ```
> altered_state_dict = torch.load(model_path, map_location=CPU)
> for i in range(12):
>     del altered_state_dict['gpt.transformer.h.' + str(i) + '.attn.bias']
>     del altered_state_dict['gpt.transformer.h.' + str(i) + '.attn.masked_bias']
> ```
> 
> And then loading the altered state dictionary into the model by:
> 
> ```
> model.load_state_dict(altered_state_dict)
> ```
> 
> ## Most likely explanation
> More specifically, in transformers: version 4.11.3 used in the original notebook, in the source code for GPT2Attention, "bias" and "masked_bias" are persistent buffers, and therefore saved to the state_dictionary: ![image](https://private-user-images.githubusercontent.com/41694147/267154401-e31a0c19-5c9f-48aa-989a-6cd70eddda91.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MzI5NTY0MTUsIm5iZiI6MTczMjk1NjExNSwicGF0aCI6Ii80MTY5NDE0Ny8yNjcxNTQ0MDEtZTMxYTBjMTktNWM5Zi00OGFhLTk4OWEtNmNkNzBlZGRkYTkxLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDExMzAlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQxMTMwVDA4NDE1NVomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTExMzQ5OWE0NzA1NzU1NjQyMmMwYzNhMTk5MmY5ZmFkZmM2NDhjN2VhZGI4YzBlNmQ3YjkwZmU5N2U2MzljYTgmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0._tptJ6H1tUru4VWrfB27aOzUvO8byAsdpWS7LgI2on8)
> 
> In the latest transformers version (version 4.33.1), in GPT2Attention, "bias" and "masked_bias" buffers are defined to have persistent=False, meaning that they are no longer a part of the model's state dictionary: ![image](https://private-user-images.githubusercontent.com/41694147/267154449-dc73b79d-8b08-47f2-a09b-c6a06721a62d.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MzI5NTY0MTUsIm5iZiI6MTczMjk1NjExNSwicGF0aCI6Ii80MTY5NDE0Ny8yNjcxNTQ0NDktZGM3M2I3OWQtOGIwOC00N2YyLWEwOWItYzZhMDY3MjFhNjJkLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDExMzAlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQxMTMwVDA4NDE1NVomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTkzMTNlNDVkMDdjYTFjYTc2ZmRkNWNjMTExZjgzNWIyOTk1MTYwNWI1NjI5NDFkOGQ3YWQ2NTg1MmUwZjIwNWImWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.XwRtYXai-fy_ey5O6BsIl7Nk4_psvq045Fa_OH6Eiq8)

thank you bro!!!!
